src/jit/morph.cpp

```diff
@@ -6940,263 +6945,236 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
+    size_t callerIntArgRegCount      = 0;
     size_t calleeFloatArgRegCount = 0;
 
     if (callee->gtCallObjp) // thisPtr
     {
-        ++calleeArgRegCount;
+        ++callerIntArgRegCount;
     }
 
-    if (callee->HasRetBufArg()) // RetBuf
-    {
-        // We don't increment calleeArgRegCount here, since it is already in callee->gtCallArgs.
-
-        // If callee has RetBuf param, caller too must have it.
-        // Otherwise go the slow route.
-        if (info.compRetBuffArg == BAD_VAR_NUM)
-        {
-            reportFastTailCallDecision("Callee has RetBuf but caller does not.", 0, 0);
-            return false;
-        }
-    }
-
-    // Count user args while tracking whether any of them is a multi-byte params
-    // that cannot be passed in a register. Note that we don't need to count
+    // Count user args while tracking whether any of them has a larger than one
+    // stack slot sized requirement. This requirement is required to support
+    // lowering the fast tail call. Which, currently only supports copying
+    // stack slot arguments which have only one stack slot.
+    //
+    // Note that we don't need to count
     // non-standard and secret params passed in registers (e.g. R10, R11) since
     // these won't contribute to out-going arg size.
-    // For each struct arg, hasMultiByteStackArgs will track if it can be passed in registers.
-    // If it cannot we will break the loop and not fastTailCall. This is an implementation limitation
-    // where the callee only is checked for non enregisterable structs.
-    // It is tracked with https://github.com/dotnet/coreclr/issues/12644.
-    bool   hasMultiByteStackArgs = false;
-    bool   hasTwoSlotSizedStruct = false;
-    bool   hasHfaArg             = false;
-    size_t nCalleeArgs           = calleeArgRegCount; // Keep track of how many args we have.
-    size_t calleeStackSize       = 0;
-    for (GenTree* args = callee->gtCallArgs; (args != nullptr); args = args->gtOp.gtOp2)
-    {
-        ++nCalleeArgs;
-        assert(args->OperIsList());
-        GenTree* argx = args->gtOp.gtOp1;
+    // 
+    // For each struct arg, determine whether the argument would have  > 1 stack 
+    // slot if on the stack If it has > 1 stack slot we will not fastTailCall. 
+    // This is an implementation limitation of LowerFastTailCall that is tracked by:
+    // https://github.com/dotnet/coreclr/issues/12644.
+
+    bool   hasLargerThanOneStackSlotSizedStruct = false;
+    bool   hasByrefStruct                       = false;
+    size_t calleeStackSize                      = 0;
 
-        if (varTypeIsStruct(argx))
+    for (unsigned index = 0; index < nCalleeArgs; ++index)
+    {
+        fgArgTabEntry* arg = argInfo->GetArgEntry(index);
+
+        if (!arg->isStruct)
         {
-            // Actual arg may be a child of a GT_COMMA. Skip over comma opers.
-            while (argx->gtOper == GT_COMMA)
+            if (arg->numSlots > 0)
```

Comment:
I wonder if there is an opportunity for generalization here where we just always add the number of integer registers, floating-point registers, and stack slots required for an argument. That would cover the arm32 reg/stack split case (arm32 currently doesn't fast tailcall, but still), cover the Linux/x64 16-byte struct integer + FP reg case, etc. And let the arg have the logic, not here. E.g.,
```
calleeStackSize += arg->StackSize();
calleeFloatArgRegCount += arg->FloatRegCount();
calleeIntArgRegCount += arg->IntRegCount();
```

Reply:
Done and great idea, this made things significantly easier to read.